Add a unit test for the "ValidBlockLibraryFunctionNameSniff" sniff

anton-vlasenko opened the pull request:
<!-- Thank you for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR aims to add a unit test for the recently introduced `ValidBlockLibraryFunctionNameSniff` sniff.
It also brings a few other minor improvements.
Fixes https://github.com/WordPress/gutenberg/issues/53731.

## Why?
It's considered best practice to write unit tests for PHPCS sniffs.
Additionally, as Gutenberg moves toward bumping the minimum supported PHP version, this package should also be updated.


## How?
<!-- How does your PR address the issue? Are there any specific implementation details to note? -->

 - [x] Add a unit test for the `ValidBlockLibraryFunctionNameSniff` sniff.
 - [x] Bump the minimum required PHP version to 7.0.
 - [x] Refine `phpcs.xml.dist` by only excluding necessary rules.
 - [x] Update WPCS to 3.0.
 - [x] Refactor `ValidBlockLibraryFunctionNameSniff::processFunctionToken()`: eliminate the need to loop through all available prefixes once the function name is confirmed as valid; this results in a minor performance improvement.
 - [x] Remove `dealerdirect/phpcodesniffer-composer-installer` from the list of dependencies per WPCS 3.0 [upgrade guide](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Upgrade-Guide-to-WordPressCS-3.0.0-for-ruleset-maintainers).

## Testing Instructions

1. Ensure that GitHub CI jobs pass.
2. Navigate to `test/php/gutenberg-coding-standards/`.
3. Run `composer update`.
4. Run `composer run check-all`.

Ensure that the PHPUnit tests in the console pass. PHPUnit should report 0 assertions, but that is expected.

`Note`: The unit tests will not work on PHP 8.0 and above. This is because the latest stable version of the `squizlabs/php_codesniffer` package [doesn't support](https://packagist.org/packages/squizlabs/php_codesniffer#3.7.2) PHPUnit 8.x. Also, PHPUnit 7.x [is not compatible](https://phpunit.de/supported-versions.html) with PHP 8.0. The unit tests have been confirmed to work on PHP 7.3.

ironprogrammer commented:
Thanks for this PR, @anton-vlasenko!

## Test Report

### Actual Results
- ✅ PHPUnit tests pass when run in the terminal (`composer run check-all` under PHP 7.4.33).

**Additional Scenarios**
- ✅ Linting PHP in the project shows that current block functions pass (`npm run lint:php`).
- ✅ Adding an invalid function name to a block gets caught by the new sniff, `Gutenberg.NamingConventions.ValidBlockLibraryFunctionName.FunctionNameInvalid` (e.g. when adding `register_block_core_imagetest()` with missing underscore after prefix). See Figure 1.

### Supplemental Artifacts
*Figure 1:*
<img width="804" alt="Invalid function name sniff result example" src="https://github.com/WordPress/gutenberg/assets/824344/181746c7-179f-4add-89f6-402ecdef8ba0">

ironprogrammer commented on test/php/gutenberg-coding-standards/Gutenberg/Sniffs/NamingConventions/ValidBlockLibraryFunctionNameSniff.php:
```diff
@@ -114,22 +114,21 @@ private function processFunctionToken( File $phpcsFile, $stackPointer ) {
 
 		$error_message = "The function name `{$function_name}()` is invalid because PHP function names in this file should start with one of the following prefixes: `"
-		                 . implode( '`, `', $allowed_function_prefixes ) . '`.';
+		. implode( '`, `', $allowed_function_prefixes ) . '`.';
```
Should each prefix listed include the required underscore? The error doesn't explicitly indicate this character is missing, as seen in the [test output shared here](https://github.com/WordPress/gutenberg/pull/53928#issuecomment-1714734896). Technically, `register_block_core_imagetest` does start with `register_block_core_image`.

anton-vlasenko replied:
Thank you for the test report, @ironprogrammer.

The reason the underscore (`_`) wasn't included is to allow function names like `register_block_core_image()`.
In my opinion, listing allowed prefixes with underscores in the error message might confuse some users.
They could mistakenly think that `register_block_core_image()` and similar function names aren't permitted.

anton-vlasenko replied:
I don't mind changing the error message text, but it should clearly indicate that:
- functions with names that are exactly equal to prefixes (**_without underscores_**) are allowed;
- functions with names that start with one of the allowed prefixes (**_including underscores_**) are also allowed.

This might be suitable for a follow-up PR, since this PR primarily focuses on adding unit tests and updating the required composer dependencies.

anton-vlasenko replied:
I've changed my mind and updated the error message. I'd appreciate your feedback, @ironprogrammer. Thank you.

ironprogrammer replied:
The updated message is more concise and works for me 👍🏻 

```
 321 | ERROR | [ ] The function name 'register_block_core_imagetest()' is invalid. In this file,
     |       |     PHP function names must either match one of the allowed prefixes exactly or
     |       |     begin with one of them, followed by an underscore. The allowed prefixes are:
     |       |     'block_core_image', 'render_block_core_image', 'register_block_core_image'.
     |       |     (Gutenberg.NamingConventions.ValidBlockLibraryFunctionName.FunctionNameInvalid)
```

anton-vlasenko replied:
@ironprogrammer Awesome! Would you approve this PR, please?

anton-vlasenko committed:
1. Add a unit test for the `ValidBlockLibraryFunctionNameSniff` sniff.
 2. Bump the minimum required PHP version to 7.0.
 3. Refine `phpcs.xml.dist` by only excluding necessary rules.
 4. Update WPCS to 3.0.
 5. Refactor `ValidBlockLibraryFunctionNameSniff::processFunctionToken()`: eliminate the need to loop through all available prefixes once the function name is confirmed as valid; this results in a minor performance improvement.
 6. Remove `dealerdirect/phpcodesniffer-composer-installer` from the list of dependencies per WPCS 3.0 [upgrade guide](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Upgrade-Guide-to-WordPressCS-3.0.0-for-ruleset-maintainers).

ironprogrammer reviewed (approved):
I re-ran tests after the recent syncs with `trunk` and it looks good, @anton-vlasenko -- I approve this PR.

anton-vlasenko commented:
Thank you, @ironprogrammer !
